pkl-commons-test/src/main/kotlin/org/pkl/commons/test/AbstractSnippetTestsEngine.kt

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.commons.test
+
+import java.nio.file.Files
+import java.nio.file.Path
+import kotlin.io.path.isRegularFile
+
+abstract class AbstractSnippetTestsEngine : InputOutputTestEngine() {
+  private val lineNumberRegex = Regex("(?m)^(( ║ )*)(\\d+) \\|")
+  private val hiddenExtensionRegex = Regex(".*[.]([^.]*)[.]pkl")
+
+  protected abstract val snippetsDir: Path
+
+  private val expectedOutputDir: Path by lazy { snippetsDir.resolve("output") }
+
+  /**
+   * Convenience for development; this selects which snippet test(s) to run. There is a
+   * (non-language-snippet) test to make sure this is `""` before commit.
+   */
+  // language=regexp
+  val selection: String = ""
+
+  protected val packageServer: PackageServer = PackageServer()
+
+  override val includedTests: List<Regex> = listOf(Regex(".*$selection\\.pkl"))
+
+  override val excludedTests: List<Regex> = listOf(Regex(".*/native/.*"))
+
+  override val inputDir: Path by lazy { snippetsDir.resolve("input") }
+
+  override val isInputFile: (Path) -> Boolean = { it.isRegularFile() }
+
+  protected tailrec fun Path.getProjectDir(): Path? =
+    if (Files.exists(this.resolve("PklProject"))) this else parent?.getProjectDir()
+
+  override fun expectedOutputFileFor(inputFile: Path): Path {
+    val relativePath = inputDir.relativize(inputFile).toString()
+    val stdoutPath =
+      if (relativePath.matches(hiddenExtensionRegex)) relativePath.dropLast(4)
+      else relativePath.dropLast(3) + "pcf"
+    return expectedOutputDir.resolve(stdoutPath)
+  }
+
+  override fun afterAll() {
+    packageServer.close()
+  }
+
+  protected fun String.stripFilePaths() = replace(snippetsDir.toString(), "/\$snippetsDir")
+
+  protected fun String.stripLineNumbers() =
+    replace(lineNumberRegex) { result ->
+      // replace line number with equivalent number of 'x' characters to keep formatting intact
+      (result.groups[1]!!.value) + "x".repeat(result.groups[3]!!.value.length) + " |"
+    }
+}
```

Comment:
I don't know if this change improves our code; this moves some fields that are only relevant to the language snippet test engine (e.g. the package server, and also `selection`, which is a quality of life thing when working on snippet tests for pkl-core.

I think let's keep it the way it was. Also, we might move the binary evaluator into pkl-core, which means the binary evaluator snippet tests will change then, making this abstraction for naught.

Reply:
It fixes dozens of IntelliJ warnings and eliminates code duplication. PackageServer is defined in commons-test and only started if needed. And yes, binary evaluator belongs into pkl-core.

Reply:
Let's just suppress those warnings for now, then; once we move it into pkl-core, this will all change again, and we'd probably end up removing this abstraction layer if it existed.

Reply:
So you’re saying that I should keep the changes that suppress the warnings and discard the remaining changes, even though they remove significant code duplication and will make it easier to move the tests into pkl-core?

Reply:
The abstraction might create more work for us later on if we eliminate `BinaryEvaluatorSnippetTests` later on if we move the binary evaluator into pkl-core.

Also, this would give us: `LanguageSnippetTestsEngine` -> `AbstractLanguageSnippetTestsEngine` -> `AbstractSnippetTestsEngine` -> `InputOutputTestEngine` -> `HierarchicalTestEngine`. That's just a lot of complexity that adds cognitive burden when trying to understand our test code.

It also moves properties that either shouldn't be moved, or don't have any meaning for the binary evaluator snippet test (`val selection: String = ""` is a property that exists for the purpose of working on the snippet tests in pkl-core).

To reduce the friction, I'd be willing to accept this PR if we moved these members into `AbstractLanguageSnippetTestsEngine`:

* `packageServer`
* `selection`
* `includedTests`
* `excludedTests`
* `Path.getProjectDir()`